File: Automatas/automata.py
import sys   # Para leer los parámetros desde la terminal

# -------------------------
# FUNCION: cargar_config
# Lee el archivo config.txt y devuelve la quíntupla (Q, Σ, δ, q0, F)
# -------------------------
def cargar_config(archivo):
    with open(archivo) as f:
        # strip() quita espacios/saltos de línea
        # ignora líneas vacías o que empiecen con "#"
        lineas = [l.strip() for l in f if l.strip() and not l.startswith("#")]
    
    # saca cada parte de la quíntupla
    estados = lineas[0].split(":")[1].split(",")     # Q = {q0,q1,q2}
    alfabeto = lineas[1].split(":")[1].split(",")    # Σ = {0,1}
    inicio = lineas[2].split(":")[1]                 # q0
    finales = lineas[3].split(":")[1].split(",")     # F = {q2}

    # δ: transiciones
    transiciones = {}
    for l in lineas[5:]:   # desde "transitions:" en adelante
        izq, der = l.split("->")     # ej: "q0,0" -> "q1"
        estado, simbolo = izq.split(",")  # separa origen y símbolo
        transiciones[(estado, simbolo)] = der  # guarda δ(origen,simbolo)=destino
    
    return estados, alfabeto, inicio, finales, transiciones

# -------------------------
# FUNCION: procesar_cadena
# Simula el AFD sobre una cadena dada
# -------------------------
def procesar_cadena(cadena, inicio, finales, transiciones, alfabeto):
    estado = inicio  # empezamos en q0
    for c in cadena:
        if c not in alfabeto:   # si símbolo no pertenece a Σ, rechaza
            return False
        if (estado, c) in transiciones:   # si hay transición definida
            estado = transiciones[(estado, c)]  # avanza al siguiente estado
        else:
            return False   # transición no definida → rechazo
    return estado in finales   # acepta si terminó en estado final

# -------------------------
# PROGRAMA PRINCIPAL
# -------------------------
def main():
    if len(sys.argv) != 3:
        print("Uso: python afd.py config.txt cadenas.txt")
        return

    # cargar quíntupla
    estados, alfabeto, inicio, finales, transiciones = cargar_config(sys.argv[1])
    
    # abrir archivo de cadenas
    with open(sys.argv[2]) as f:
        for linea in f:
            cadena = linea.strip()
            if not cadena: continue  # ignorar vacías
            aceptada = procesar_cadena(cadena, inicio, finales, transiciones, alfabeto)
            # salida
            print("acepta" if aceptada else "NO acepta")

if __name__ == "__main__":
    main()

File: Automatas/automata.c
#include <stdio.h>
#include <string.h>
#include <stdlib.h>

#define MAX 100

typedef struct {
    char origen[10];
    char simbolo;
    char destino[10];
} Transicion;

int main(int argc, char *argv[]) {
    if (argc != 3) {
        printf("Uso: %s config.txt cadenas.txt\n", argv[0]);
        return 1;
    }

    FILE *fconf = fopen(argv[1], "r");
    FILE *fcad = fopen(argv[2], "r");
    if (!fconf || !fcad) {
        printf("Error abriendo archivos\n");
        return 1;
    }

    char linea[100];
    char estados[MAX][10], finales[MAX][10], inicio[10];
    int num_finales = 0, num_trans = 0;
    Transicion trans[MAX];

    while (fgets(linea, sizeof(linea), fconf)) {
        if (strncmp(linea, "start:", 6) == 0) {
            sscanf(linea, "start:%s", inicio);
        } else if (strncmp(linea, "finals:", 7) == 0) {
            char *tok = strtok(linea+7, ",\n");
            while (tok) {
                strcpy(finales[num_finales++], tok);
                tok = strtok(NULL, ",\n");
            }
        } else if (strstr(linea, "->")) {
            char izq[20], der[20];
            sscanf(linea, "%[^-]->%s", izq, der);
            char *coma = strchr(izq, ',');
            *coma = '\0';
            strcpy(trans[num_trans].origen, izq);
            trans[num_trans].simbolo = *(coma+1);
            strcpy(trans[num_trans].destino, der);
            num_trans++;
        }
    }

    while (fgets(linea, sizeof(linea), fcad)) {
        char cadena[100]; strcpy(cadena, linea);
        cadena[strcspn(cadena, "\n")] = 0;
        if (strlen(cadena) == 0) continue;

        char estado[10]; strcpy(estado, inicio);
        for (int i = 0; i < strlen(cadena); i++) {
            char simbolo = cadena[i];
            int encontrado = 0;
            for (int j = 0; j < num_trans; j++) {
                if (strcmp(trans[j].origen, estado) == 0 && trans[j].simbolo == simbolo) {
                    strcpy(estado, trans[j].destino);
                    encontrado = 1;
                    break;
                }
            }
            if (!encontrado) { strcpy(estado, ""); break; }
        }

        int aceptada = 0;
        for (int i = 0; i < num_finales; i++) {
            if (strcmp(estado, finales[i]) == 0) aceptada = 1;
        }

       if (aceptada)
            printf("acepta\n");
        else
            printf("NO acepta\n");
    }

    fclose(fconf);
    fclose(fcad);
    return 0;
}

